### PR TITLE
#21955: Fixed the case when identity flow appended to Source combining materialized values

### LINKED ISSUE #21955: Materialized value combination does not work with identity Flows
reproducer:

```
val ((pub, mat), sub) = TestSource.probe[Int].viaMat(Flow[Int])(Keep.both).toMat(TestSink.probe)(Keep.both).run()
```

Throws ClassCastException. The reason is simple, the DSL detects the identity Flow and simply converts viaMat(Keep.both) to a NOP. This is correct for the graph, but not for the materialized value which needs the NotUsed value coming from the empty Flow.

### PR DESCRIPTION
**Pull request by drewhk**
Fixes #21955

### CODE REVIEW

**Comment by drewhk** [comment]
Needs backport

**Review by ktoso — APPROVED** [review]
LGTM

**Review by 2m — APPROVED** [review]

**Comment by drewhk** [comment]
for the record, this is correctly handled in Flow already, and Sink does not have a via method.

**Review by patriknw — APPROVED** [review]
LGTM

File: akka-stream-tests/src/test/scala/akka/stream/scaladsl/GraphMatValueSpec.scala [hunk 1]
```diff
@@ -227,5 +227,12 @@ class GraphMatValueSpec extends StreamSpec {
       matValue should ===(NotUsed)
 
     }
+
+    "not ignore materialized value of indentity flow which is optimized away" in {
+      implicit val mat = ActorMaterializer(ActorMaterializerSettings(system).withAutoFusing(false))
+      val (m1, m2) = Source.single(1).viaMat(Flow[Int])(Keep.both).to(Sink.ignore).run()
```
Comment (viktorklang) [inline akka-stream-tests/src/test/scala/akka/stream/scaladsl/GraphMatValueSpec.scala]: `val (NotUsed, NotUsed) = Source.single(1).viaMat(Flow[Int])(Keep.both).to(Sink.ignore).run()`

No assertions needed :)

File: akka-stream/src/main/scala/akka/stream/scaladsl/Source.scala [hunk 2]
```diff
@@ -43,7 +43,8 @@ final class Source[+Out, +Mat](override val module: Module)
   override def via[T, Mat2](flow: Graph[FlowShape[Out, T], Mat2]): Repr[T] = viaMat(flow)(Keep.left)
 
   override def viaMat[T, Mat2, Mat3](flow: Graph[FlowShape[Out, T], Mat2])(combine: (Mat, Mat2) ⇒ Mat3): Source[T, Mat3] = {
-    if (flow.module eq GraphStages.Identity.module) this.asInstanceOf[Source[T, Mat3]]
+    if (flow.module eq GraphStages.Identity.module)
+      this.mapMaterializedValue(combine(_, NotUsed.asInstanceOf[Mat2])).asInstanceOf[Source[T, Mat3]]
```
Comment (viktorklang) [inline akka-stream/src/main/scala/akka/stream/scaladsl/Source.scala]: Is this the best we can do?
check if the matfun is Keep.left?

Comment (drewhk) [inline akka-stream/src/main/scala/akka/stream/scaladsl/Source.scala]: I can check both Keep.left and Keep.right, sure.

File: akka-stream-tests/src/test/scala/akka/stream/scaladsl/GraphMatValueSpec.scala (hunk 1 above)
Comment (drewhk) [inline akka-stream-tests/src/test/scala/akka/stream/scaladsl/GraphMatValueSpec.scala]: lol, right :)

Comment (drewhk) [inline akka-stream-tests/src/test/scala/akka/stream/scaladsl/GraphMatValueSpec.scala]: I keep the assertions though as they provide better error messages and the test is more readable.

**Comment by drewhk** [comment]
I don't see any actual failure in the validation run. The only thing I found is

```
Build step 'Execute a set of scripts' changed build result to FAILURE
Build step 'Execute a set of scripts' marked build as failure
```

at the end, but I don't know what that means.

**Comment by 2m** [comment]
Just before that, there are some
```
Permission denied, please try again.
Permission denied, please try again.
Permission denied (publickey,password).
```

which is part of multi-node test cleanup. It looks everything else completed without errors.

**Comment by drewhk** [comment]
PLS BUILD

**Comment by drewhk** [comment]
unrelated failure

**Pull request merged**